controllers/databaseengine_controller.go

```diff
@@ -181,135 +175,147 @@ func (r *DatabaseEngineReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+		return err
+	}
+	if idx := slices.IndexFunc(ipList.Items, func(ip opfwv1alpha1.InstallPlan) bool {
+		return slices.Contains(ip.Spec.ClusterServiceVersionNames, newCSV.GetName())
+	}); idx > 0 {
+		installPlan = &ipList.Items[idx]
+	} else {
+		return errInstallPlanNotFound
+	}
+
+	if csv.Status.Phase == opfwv1alpha1.CSVPhaseReplacing {
+		dbEngine.Status.State = everestv1alpha1.DBEngineStateUpgrading
+		if dbEngine.Status.OperatorUpgrade == nil {
+			_, targetVersion := parseOperatorCSVName(newCSV.GetName())
+			now := metav1.Now()
+			dbEngine.Status.OperatorUpgrade = &everestv1alpha1.OperatorUpgradeStatus{
+				OperatorUpgrade: everestv1alpha1.OperatorUpgrade{
+					TargetVersion: targetVersion,
+					InstallPlanRef: corev1.LocalObjectReference{
+						Name: installPlan.GetName(),
+					},
+				},
+				Phase:     everestv1alpha1.UpgradePhaseStarted,
+				StartedAt: &now,
+				Message:   "Upgrading operator to version " + targetVersion,
+			}
 		}
-		// Upgrade is complete, remove the annotation and mark upgrade as complete.
-		dbEngine.Status.OperatorUpgrade.Phase = everestv1alpha1.UpgradePhaseCompleted
-		// These annotations are set by the everest backend.
-		delete(annotations, everestv1alpha1.DatabaseOperatorUpgradeAnnotation)
-		delete(annotations, everestv1alpha1.DatabaseOperatorUpgradeLockAnnotation)
-		dbEngine.SetAnnotations(annotations)
-		return false, r.Update(ctx, dbEngine)
 	}
 
-	if installPlan.Status.Phase == opfwv1alpha1.InstallPlanPhaseFailed {
+	if newCSV.Status.Phase == opfwv1alpha1.CSVPhaseFailed {
 		dbEngine.Status.OperatorUpgrade.Phase = everestv1alpha1.UpgradePhaseFailed
-		return false, nil
 	}
-	return false, nil
+	return nil
 }
 
-func (r *DatabaseEngineReconciler) listPendingOperatorUpgrades(
+func (r *DatabaseEngineReconciler) tryUnlockDBEngine(
 	ctx context.Context,
 	dbEngine *everestv1alpha1.DatabaseEngine,
-) ([]everestv1alpha1.OperatorUpgrade, error) {
-	// If OLM is not installed, we cannot check for pending upgrades.
-	if !r.isOLMInstalled(ctx) {
-		return nil, nil
+) error {
+	logger := log.FromContext(ctx)
+	annotations := dbEngine.GetAnnotations()
+	if annotations == nil {
+		return nil
 	}
-	// We need some version to be reported first.
-	currentVersion := dbEngine.Status.OperatorVersion
-	if currentVersion == "" {
-		return nil, nil
+	lockedAtStr, ok := annotations[everestv1alpha1.DatabaseOperatorUpgradeLockAnnotation]
+	if !ok {
+		return nil
+	}
+	lockedAt, err := time.Parse(time.RFC3339, lockedAtStr)
+	if err != nil {
+		return errors.Join(err, errors.New("cannot parse the value of 'everest.percona.com/upgrade-lock' annotation"))
+	}
+	// Check if it has exceeded the upgrade timeout.
+	if time.Now().After(lockedAt.Add(upgradeTimeout)) {
+		logger.Info("databaseengine upgrade timeout, force unlock triggered")
+		delete(annotations, everestv1alpha1.DatabaseOperatorUpgradeLockAnnotation)
+		dbEngine.SetAnnotations(annotations)
+		return r.Update(ctx, dbEngine)
```

Comment:
Should we invert the condition and return nil to avoid duplicating the delete annotation logic found at the end of this func?
I'm not sure this is possible though, due to the rest of the conditions checked below.

Reply:
Done, I moved the unlock logic into a separate function variable to re-use it..